docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+| Initializing  | R (Restoration)            | This state denotes that backup-restore container in etcd-member pod has started initialization. Sub-State `R` which is an abbreviation for `Restoration` denotes that DB validation failed and now backup-restore has commenced restoration of etcd DB from the backup (comprising of full snapshot and delta-snapshots). An etcd-member will transition to this sub-state only when it is part of a single-node etcd-cluster. |
+| Starting (SI) | PL (PendingLearner)        | An etcd-member can transition from `Initializing` state to `PendingLearner` state. In this state backup-restore container will optionally delete any existing etcd data directory and then attempts to add its peer etcd-member process as a learner. Since there can be only one learner at a time in an etcd cluster, an etcd-member could be in this state for some time till its request to get added as a learner is accepted. |
+| Starting (SI) | Learner                    | When backup-restore is successfully able to add its peer etcd-member process as a `Learner`. In this state the etcd-member process will start its DB sync from an etcd leader. |
+| Started (Sd)  | Follower                   | A follower is a voting raft member. A `Learner` etcd-member will get promoted to a `Follower` once its DB is in sync with the leader. It could also become a follower if during a re-election it loses leadership and transitions from being a `Leader` to `Follower`. |
+| Started (Sd)  | Leader                     | A leader is an etcd-member which will handle all client write requests and linearizable read requests. A member could transition to being a `Leader` from an existing `Follower` role due to winning a leader election or for a single node etcd cluster it directly transitions from `Initializing` state to `Leader` state as there is no other member. |
+
+In the following sub-sections, the state transitions are categorized into several flows making it easier to grasp the different transitions.
+
+#### Top Level State Transitions
+
+Following DFA represents top level state transitions (without any representation of sub-states). As described in the table above there are 4 top level states:
+
+* `New`- this is a start state for all newly created etcd-members
+
+* `Initializing` - In this state backup-restore will perform pre-requisite actions before it triggers the start of an etcd process. DB validation and optionally restoration is done in this state. Possible sub-states are: `DBValidationSanity`, `DBValidationFully` and `Restoration`
+
+* `Starting` - Once the optional initialization is done backup-restore will trigger the start of an etcd process. It can either directly go to `Learner` sub-state or wait for getting added as a learner and therefore be in `PendingLearner` sub-state.
+
+* `Started` - In this state the etcd-member is a full voting member. It can either be in `Leader` or `Follower` sub-states.
```

Comment:
WDYT of `Running` instead?

Reply:
We chose `Started` as it was a natural succession state after `Starting` but there is no hard and fast rule. `Running` is semantically same as `Started` and could be considered.

Reply:
Nice. My point is that it's visually nicer to tell the difference:
- `starting` - `started`
vs.
- `starting` - `running`

Reply:
Are we going to update the ETCD member states even for the sub-states? I think, we can cut those substates and only keep Two main states, named Initializing and Running.

Reply:
what is the rationale?

Reply:
As discussed during the meeting on Wed 23 Aug:
For now we will retain the same set of state and sub-states as described in this proposal. Also see [comment](https://github.com/gardener/etcd-druid/pull/658/files#r1303818726)